### PR TITLE
SD-4358 OpsCenter changes to support Sundeck account creation using Snowflake SSO

### PR DESCRIPTION
**Pull request by scgkiran**

### CODE REVIEW

File: app/ui/Home.py [hunk 1]
```diff
@@ -32,3 +32,10 @@
     An overview dashboard will arrive here soon!
     """
     )
+    if config.has_tenant_url():
+        tenant_url = config.get_tenant_url()
+        st.markdown(
+            f"""
+                [Click here]({tenant_url}) to visit your Sundeck account.
```
Comment (pravindra) [inline app/ui/Home.py]: `[Go to my sundeck account]({tenant_url})`

File: app/ui/setup.py [hunk 2]
```diff
@@ -109,49 +87,84 @@ def expander(num: int, title: str, finished: bool) -> st.expander:
...
+        )
 
+        st.markdown(
+            """
+                #### Use one of the two options to create a free Sundeck account
+            """
+        )
+
+        option1, option2 = st.tabs(
+            ["Option 1 Signup with Snowflake SSO", "Option 2 Signup using email-Id"]
```
Comment (pravindra) [inline app/ui/setup.py]: can you put a colon after the number (1 & 2) ?

Option 1 : Signup with Snowflake SSO

File: app/ui/setup.py [hunk 3]
```diff
@@ -175,19 +188,120 @@ def sndk_url(account: str, user: str, region: str) -> str:
...
+    st.markdown(
+        """
+        #### Sundeck-signup using Snowflake-SSO. Follow these steps to enable notifications:
+        1. Prepare to create a Sundeck account. This is needed to for step 2.
+
+        """
     )
-    return (
+
+    if st.button("Prepare to Enable Notifications", key="create_api_integration"):
```
Comment (pravindra) [inline app/ui/setup.py]: Enable Sundeck api integration

File: app/ui/setup.py [hunk 4]
```diff
@@ -175,19 +188,120 @@ def sndk_url(account: str, user: str, region: str) -> str:
...
+    oauth_issue_refresh_tokens=true
+    oauth_refresh_token_validity=86400
+    pre_authorized_roles_list = ('PUBLIC');
+"""
+
+
+def generate_code_to_register_tenant(db: str, security_integration_name: str) -> str:
+    return f"""
+let SecurityIntegration:='{security_integration_name}';
+let oauth_info variant := (parse_json(SYSTEM$SHOW_OAUTH_CLIENT_SECRETS(:SecurityIntegration)));
```
Comment (pravindra) [inline app/ui/setup.py]: remove line 287, update 288 to

`let oauth_info variant := (parse_json(SYSTEM$SHOW_OAUTH_CLIENT_SECRETS({security_integration_name})`

File: app/ui/setup.py [hunk 5]
```diff
@@ -175,19 +188,120 @@ def sndk_url(account: str, user: str, region: str) -> str:
...
+let tenantInfo object := {db}.tools.registertenant(:oauth_info:OAUTH_CLIENT_ID, :oauth_info:OAUTH_CLIENT_SECRET);
+CALL {db}.admin.setup_sundeck_tenant_url(:tenantInfo:sundeckTenantUrl, :tenantInfo:sundeckUdfToken);
+return OBJECT_CONSTRUCT('Sundeck Account ', tenantInfo:sundeckTenantUrl);
```
Comment (pravindra) [inline app/ui/setup.py]: replace line 251 with below to get a clickable URL

```
let rs resultset := (select 'Go to Sundeck UI' As msg, :tenantInfo:sundeckTenantUrl::string as url);
return table(rs);
```

File: bootstrap/021_external_functions.sql [hunk 6]
```diff
@@ -121,17 +121,54 @@ BEGIN
...
-    execute immediate 'create or replace function internal.get_tenant_url() returns string as \'\\\'' || url || '\\\'\';';
-END;
 
 create or replace procedure internal.setup_sundeck_token(url string, token string) RETURNS STRING LANGUAGE SQL AS
 BEGIN
     execute immediate 'create or replace function internal.get_ef_url() returns string as \'\\\'' || url || '\\\'\';';
     execute immediate 'create or replace function internal.get_ef_token() returns string as \'\\\'' || token || '\\\'\';';
 END;
 
+create function if not exists internal.ef_registertenant(request object)
```
Comment (pravindra) [inline bootstrap/021_external_functions.sql]: this name looks ugly. can you pls add an underscore ?

`create function if not exists internal.ef_register_tenant(request object)`

File: bootstrap/021_external_functions.sql [hunk 7]
```diff
@@ -141,21 +178,21 @@ BEGIN
             create or replace external function internal.ef_qlike(request object)
             returns object
             context_headers = (CURRENT_ACCOUNT, CURRENT_USER, CURRENT_ROLE, CURRENT_DATABASE, CURRENT_SCHEMA)
-            api_integration = reference(\'opscenter_api_integration\')
+            api_integration = reference(\'opscenter_api_integration_ad1\')
```
Comment (pravindra) [inline bootstrap/021_external_functions.sql]: what is ad1 ?

**Comment by anshuldata** [comment]
test comment

File: app/ui/setup.py [hunk 8]
```diff
@@ -284,11 +283,15 @@ def generate_code_to_create_security_integration(
...
-let SecurityIntegration:='{security_integration_name}';
-let oauth_info variant := (parse_json(SYSTEM$SHOW_OAUTH_CLIENT_SECRETS(:SecurityIntegration)));
+let oauth_info variant := (parse_json(SYSTEM$SHOW_OAUTH_CLIENT_SECRETS('{security_integration_name}')));
 let tenantInfo object := {db}.tools.registertenant(:oauth_info:OAUTH_CLIENT_ID, :oauth_info:OAUTH_CLIENT_SECRET);
 CALL {db}.admin.setup_sundeck_tenant_url(:tenantInfo:sundeckTenantUrl, :tenantInfo:sundeckUdfToken);
-return OBJECT_CONSTRUCT('Sundeck Account ', tenantInfo:sundeckTenantUrl);
+
+let rs resultset := (select 'Go to Sundeck UI' As msg,
+    :tenantInfo:sundeckTenantUrl::string as url,
+    :tenantInfo:sundeckUdfToken::string as token);
```
Comment (pravindra) [inline app/ui/setup.py]: can we not show the token ? it's an internal detail that the customer doesn't need to be aware of

File: app/ui/Home.py (hunk 1 above)
Comment (scgkiran) [reply app/ui/Home.py]: done

File: app/ui/setup.py (hunk 2 above)
Comment (scgkiran) [reply app/ui/setup.py]: done

File: app/ui/setup.py (hunk 8 above)
Comment (scgkiran) [reply app/ui/setup.py]: done

File: bootstrap/021_external_functions.sql (hunk 6 above)
Comment (scgkiran) [reply bootstrap/021_external_functions.sql]: done

File: bootstrap/021_external_functions.sql (hunk 7 above)
Comment (scgkiran) [reply bootstrap/021_external_functions.sql]: changed it to 
opscenter_api_integration
opscenter_api_integration_sso

File: app/ui/setup.py (hunk 3 above)
Comment (scgkiran) [reply app/ui/setup.py]: done

Comment (scgkiran) [reply app/ui/setup.py]: done

File: app/ui/setup.py (hunk 4 above)
Comment (scgkiran) [reply app/ui/setup.py]: done

**Review by pravindra — APPROVED** [review]
lgtm !

**Review by anshuldata — APPROVED** [review]
lgtm

**Pull request merged**